File: cmd/aistat/usage.go
package main

import (
	"context"
	"flag"
	"fmt"
	"io"
	"os"
	"os/signal"
	"slices"
	"strings"
	"syscall"

	"github.com/drogers0/aistat/v2/internal/httpx"
	"github.com/drogers0/aistat/v2/internal/orchestrate"
	"github.com/drogers0/aistat/v2/internal/providers"
	"github.com/drogers0/aistat/v2/internal/render"
)

// runUsage runs the `usage` subcommand: fetch and render provider limits.
// args contains everything after the "usage" subcommand token (or all original
// args when invoked with no subcommand). An optional first positional argument
// names a single provider to query; all other arguments are flags.
func runUsage(args []string, stdout, stderr io.Writer, g globals) int {
	fs := flag.NewFlagSet("usage", flag.ContinueOnError)
	fs.SetOutput(io.Discard)
	fs.Usage = func() {}
	registerGlobalFlags(fs, &g)
	fakeFn := registerFakeMode(fs)
	var refresh bool
	fs.BoolVar(&refresh, "refresh", false, "")

	// First pass: parse any leading flags before the optional provider positional.
	if err := fs.Parse(args); err != nil {
		fmt.Fprintln(stderr, err.Error())
		return int(orchestrate.StatusUsageError)
	}

	// After --help/--version check (they may appear after "usage" token).
	if handled, code := handleGlobals(g, stdout); handled {
		return code
	}

	// Extract the optional provider positional (first non-flag arg from first pass).
	var service string
	tail := fs.Args()
	if len(tail) > 0 {
		service = tail[0]
		tail = tail[1:]
	}

	// Second pass: parse any trailing flags after the optional provider.
	// Always call Parse (even with empty tail) so fs.Args() is reset and
	// fs.NArg() correctly reflects only unconsumed positionals from this pass.
	if err := fs.Parse(tail); err != nil {
		fmt.Fprintln(stderr, err.Error())
		return int(orchestrate.StatusUsageError)
	}

	// After --help/--version that may appear after the provider.
	if handled, code := handleGlobals(g, stdout); handled {
		return code
	}

	if fs.NArg() > 0 {
		fmt.Fprintf(stderr, "unexpected positional argument: %s\n", fs.Arg(0))
		return int(orchestrate.StatusUsageError)
	}

	if service != "" && !slices.Contains(providers.KnownProviderIDs, service) {
		fmt.Fprintf(stderr, "usage %s: provider must be one of %s\n",
			service, strings.Join(providers.KnownProviderIDs, ", "))
		return int(orchestrate.StatusUsageError)
	}

	requested := selectedProviders(service)

	serialStderr := httpx.NewConcurrencySafeWriter(stderr)
	chosen, orchDebug := buildProviders(serialStderr, g.Debug, refresh, fakeFn)

	ctx, stop := signal.NotifyContext(context.Background(), os.Interrupt, syscall.SIGTERM)
	defer stop()

	report, status := orchestrate.Run(ctx, requested, chosen, orchestrate.Options{Debug: orchDebug})

	var renderErr error
	if g.Human {
		renderErr = render.Text(stdout, report, requested)
	} else {
		renderErr = render.JSON(stdout, report)
	}
	if renderErr != nil {
		fmt.Fprintln(stderr, renderErr.Error())
		return int(orchestrate.StatusRenderError)
	}
	return int(status)
}

// selectedProviders returns the provider ID list to query. When service is
// empty, all known providers are requested.
func selectedProviders(service string) []string {
	if service == "" {
		return append([]string(nil), providers.KnownProviderIDs...)
	}
	return []string{service}
}

File: src/cli/usage.rs
use crate::orchestrate::{run, ExitStatus, RunOptions};
use crate::providers::KNOWN_PROVIDER_IDS;
use crate::render::{json::render_json, text::render_text};
use std::io::{self};

pub struct UsageArgs {
    pub provider: Option<String>,
    pub refresh: bool,
    pub human: bool,
    pub debug: bool,
}

pub fn run_usage(args: UsageArgs) -> i32 {
    if let Some(ref p) = args.provider {
        if !KNOWN_PROVIDER_IDS.contains(&p.as_str()) {
            eprintln!(
                "usage {}: provider must be one of {}",
                p,
                KNOWN_PROVIDER_IDS.join(", ")
            );
            return ExitStatus::UsageError as i32;
        }
    }

    let requested: Vec<&str> = match &args.provider {
        Some(p) => vec![p.as_str()],
        None => KNOWN_PROVIDER_IDS.to_vec(),
    };

    let providers = crate::cli::registry::real_providers(args.refresh, args.debug);

    let (report, status) = run(
        &requested,
        &providers,
        RunOptions { debug: args.debug },
    );

    let stdout = io::stdout();
    let mut out = stdout.lock();

    let render_err = if args.human {
        render_text(&mut out, &report, &requested)
    } else {
        render_json(&mut out, &report)
    };

    if let Err(e) = render_err {
        eprintln!("{}", e);
        return ExitStatus::RenderError as i32;
    }

    status as i32
}
